internal/jobcontainers/path_test.go

```diff
@@ -31,52 +33,159 @@ func TestSearchPath(t *testing.T) {
+func TestGetApplicationNamePing(t *testing.T) {
 	expected := "C:\\WINDOWS\\system32\\ping.exe"
 
 	cwd, err := os.Getwd()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	path, _, err := getApplicationName("ping", cwd, os.Getenv("PATH"))
+	pathEnv := os.Getenv("PATH")
+
+	path, _, err := getApplicationName("ping", cwd, pathEnv)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertStr(t, expected, path)
+
+	path, _, err = getApplicationName("./ping", cwd, pathEnv)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assertStr(t, expected, path)
 
-	path, _, err = getApplicationName("./ping", cwd, os.Getenv("PATH"))
+	path, _, err = getApplicationName(".\\ping", cwd, pathEnv)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assertStr(t, expected, path)
 
-	path, _, err = getApplicationName(".\\ping", cwd, os.Getenv("PATH"))
+	// Test relative path with different cwd
+	newCwd := `C:\Windows\`
+	_, _, err = getApplicationName("./system32/ping", newCwd, pathEnv)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assertStr(t, expected, path)
 
+	pingWithCmd := "cmd /c ping 127.0.0.1"
+	path, cmdLine, err := getApplicationName("cmd /c ping 127.0.0.1", cwd, pathEnv)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertStr(t, cmdLine, pingWithCmd)
+	assertStr(t, "C:\\windows\\system32\\cmd.exe", path)
+
 	// Test that we only find the first element of the commandline if the binary exists.
-	path, _, err = getApplicationName("ping test", cwd, os.Getenv("PATH"))
+	path, _, err = getApplicationName("ping test", cwd, pathEnv)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assertStr(t, expected, path)
 
 	// Test quoted application name with an argument afterwards.
-	path, cmdLine, err := getApplicationName("\"ping\" 127.0.0.1", cwd, os.Getenv("PATH"))
+	path, cmdLine, err = getApplicationName("\"ping\" 127.0.0.1", cwd, pathEnv)
 	if err != nil {
 		t.Fatal(err)
 	}
 	assertStr(t, expected, path)
 
-	args := splitArgs(cmdLine)
 	cmd := &exec.Cmd{
 		Path: path,
-		Args: args,
+		Args: splitArgs(cmdLine),
 	}
 	if err := cmd.Run(); err != nil {
 		t.Fatal(err)
 	}
 }
+
+func TestGetApplicationNameRandomBinary(t *testing.T) {
+	pathEnv := os.Getenv("PATH")
+
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	// Create fake executables in a temporary directory to use for the below tests.
+	testExe := filepath.Join(tempDir, "test.exe")
+	_, err = os.Create(testExe)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	test2Exe := filepath.Join(tempDir, "test 2.exe")
+	_, err = os.Create(test2Exe)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	exeWithSpace := filepath.Join(tempDir, "exe with space.exe")
+	_, err = os.Create(exeWithSpace)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// See if we can successfully find "exe with space.exe" with no quoting, it should first try "exe.exe", then "exe with.exe" and then finally
+	// "exe with space.exe"
+	path, _, err := getApplicationName("exe with space.exe", filepath.Dir(testExe), os.Getenv("PATH"))
```

Comment:
Can we turn all these test cases into a loop over structs with `t.Run`?

Reply:
Sure, they're getting a bit unwieldy..